src/codeu/chat/util/Timeline.java

```diff
@@ -0,0 +1,146 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package codeu.chat.util;
+
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.PriorityBlockingQueue;
+
+// TIMELINE
+//
+// The timeline is a time ordered collection of executable units. This is used
+// when work needs to be ordered by time. The timeline manages its own threads
+// and there is no way to know outside of the code that is executed when the
+// code has been executed.
+public final class Timeline {
+
+  private static final class Event implements Comparable<Event> {
+
+    public final long time;
+    public final Runnable callback;
+
+    public Event(long time, Runnable callback) {
+      this.time = time;
+      this.callback = callback;
+    }
+
+    @Override
+    public int compareTo(Event other) {
+      return Long.compare(time, other.time);
+    }
+  }
+
+  private final BlockingQueue<Event> backlog = new PriorityBlockingQueue<>();
+  private final BlockingQueue<Runnable> todo = new LinkedBlockingQueue<>();
+
+  private boolean running = true;
+
+  // This thread is used to track the time of events and moves events from the
+  // "backlog" queue to the "todo" queue when it is time to execute. They are
+  // seperated to allow this thread to be safely interrupted.
+  private final Thread scheduler = new Thread() {
+    @Override
+    public void run() {
+      while (running) {
+        final long now = System.currentTimeMillis();
+        final Event next = backlog.poll();
+        try {
+          if (next == null) {
+            Thread.sleep(10000);
```

Comment:
But it won't. When a new item is added to the queue, the thread is interrupted which will force it to awake from a sleep. I would have rather it sleep indefinitely (but that was harder to do) because it will get woken from the interrupt.